### PR TITLE
dev/core#1895 fix first/last name adv search

### PR DESCRIPTION
**Pull request by lcdservices**
Overview
----------------------------------------
The ability to search first/last name was recently added to core. The primary intent was to properly handle advanced searches that originate from quicksearch first/last name filters. While that was working, a first/last name search that originated directly from an advanced search was not working. This fixes the issue.

Before
----------------------------------------
Go to advanced search, expose the first/last name fields, enter criteria and initiate the search. Result = all records returned and the filter value was not preserved.

After
----------------------------------------
Conduct search and it returns results as expected.

Technical Details
----------------------------------------
There were a couple issues with the original patch that needed to be addressed. The first/last name fields needed to be constructed in the form explicitly, rather than through the metadata function, as they were not handled properly by the form via the previous method. And the js that toggles between the sort_name and first_name/last_name fields needed to use a method other than show()/hide() as the resulting display:none meant that the form values were lost on reload.

### CODE REVIEW

File: CRM/Contact/Form/Search/Criteria.php [hunk 1]
```diff
@@ -28,6 +28,9 @@ public static function basic(&$form) {
     self::setBasicSearchFields($form);
     $form->addElement('hidden', 'hidden_basic', 1);
 
+    $form->add('text', 'first_name', ts('First Name'), CRM_Core_DAO::getAttribute('CRM_Contact_DAO_Contact', 'first_name'));
+    $form->add('text', 'last_name', ts('Last Name'), CRM_Core_DAO::getAttribute('CRM_Contact_DAO_Contact', 'last_name'));
+
```
Comment (seamuslee001) [inline CRM/Contact/Form/Search/Criteria.php:33]: is this needed @lcdservices

**Comment by monishdeb** [comment]
This patch exposes a bug in metadata function where it shows that if we explicitly declare a form element in a search form, it works and also retain the keyword on form reload. But when we use the metadata function to add search fields first and last name it doesn't work. Reason why in the patch, @lcdservices has tried to move the fields declaration from metadata to search form. I am trying to figure out the cause why its only happening with these two fields not with sort_name or other basic contact fields.

**Comment by monishdeb** [comment]
@lcdservices @seamuslee001 I have made the fix and updated the PR. First and last name is now working on advance search and also with quick search. Please have a look.

**Comment by lcdservices** [comment]
This is working. I'm comfortable with the code. It approaches the issue in a slightly different way than I had proposed, but addresses the same issues I had run into.

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
This test is very form layer so I don't think it needs a test. I'm comfortable with the approach and the face @lcdservices has r-run it. My one question preventing me merging this now is that I wonder if it shouldn't go against the rc.

Based on that I'm going to add merge-ready. One way or another we should merge before the next rc is cut

**Comment by eileenmcnaughton** [comment]
Test Result (1 failure / +1)api_v3_JobTest.testNoErrorOnOdd
--



presumably unrelated but not familiar

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
@lcdservices @monishdeb same error - I think it must relate

**Comment by monishdeb** [comment]
Looking at these test failure

**Comment by monishdeb** [comment]
Not sure what triggers this API failure but on local .. every time I get this error which is different from the error listed in test build: 
```
Monishs-MacBook-Pro:civicrm monish$ phpunit5 tests/phpunit/api/v3/JobTest.php 
Parsing schema description /Users/monish/src/civicrm-core/xml/schema/Schema.xml
Extracting database information
Extracting table information
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

.
Installing d7civimast_2yhrl database
..............................................................  63 / 203 ( 31%)
...................................................E........... 126 / 203 ( 62%)
............................................................... 189 / 203 ( 93%)
...........
<script type="text/javascript">
  CRM.$(function($) {
    $('#birth_date_from, #birth_date_to').attr({startOffset: '200', endoffset: '0'});
  });
</script>


      <div class="crm-block crm-form-block crm-report-field-form-block">
```

File: templates/CRM/Contact/Form/Search/Criteria/Basic.tpl [hunk 2]
```diff
@@ -10,7 +10,7 @@
 <div class="advanced-search-fields basic-fields form-layout">
   {foreach from=$basicSearchFields item=fieldSpec}
     {assign var=field value=$form[$fieldSpec.name]}
-    {if $field}
+    {if $field && !in_array($fieldSpec.name, array('first_name', 'last_name'))}
```
Comment (eileenmcnaughton) [inline templates/CRM/Contact/Form/Search/Criteria/Basic.tpl:13]: Do we suppress the fields here?

**Comment by eileenmcnaughton** [comment]
@seamuslee001 I know you were digging into this - it looks OK to me - if it does to you please merge

File: templates/CRM/Contact/Form/Search/Criteria/Basic.tpl (hunk 2 above)
Comment (monishdeb) [reply templates/CRM/Contact/Form/Search/Criteria/Basic.tpl:13]: @eileenmcnaughton no. As in sort_name.tpl we are already rendering these two fields,  reason why we are preventing first_name and last_name to be added as a separate field in advance search form.

**Comment by eileenmcnaughton** [comment]
@seamuslee001 we should put against the rc now? Where is this at from your pov?

**Comment by seamuslee001** [comment]
I'm good with this updated the base branch and changed to MOP

**Comment by monishdeb** [comment]
Thanks @eileenmcnaughton @seamuslee001

**Pull request merged**